### DISCUSSION TITLE
`$` is not defined

### DISCUSSION THREAD

**Discussion by lovelyJason** [vercel/next.js#65032] [Help] [answered]
### Summary

看图片
![image](https://github.com/vercel/next.js/assets/50656459/7df7f74f-0ed8-49a1-a8ac-27ebe5e14aae)

- src/app/layout.tsx
```tsx
import Script from "next/script";

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body>
        <Script strategy="beforeInteractive" src="/js/jquery-3.4.1.min.js" />
        <div id="app">
   
          {children}
        </div>
      </body>
    </html>
  );
}
```
Isn't jquery here a global reference?

And except for Layout.tsx, all my pages are added with 'use client' '. Even if I added 'use client' at the top of `src/app.Layout.tsx`, I also reported this wrong, why?Why did you report this fault and how to rsolve it。

## webpack.ProvidePlugin

And when I use webpack.ProvidePlugin like this:
```js
/** @type {import('next').NextConfig} */
import { error } from "node:console";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import fastSafeStringify from "fast-safe-stringify";

import {
  PHASE_DEVELOPMENT_SERVER,
  PHASE_PRODUCTION_BUILD,
} from "next/dist/shared/lib/constants.js";
import CopyWebpackPlugin from "copy-webpack-plugin";
import webpack from "webpack";
import { merge } from "webpack-merge"; // 不同于lodash.merge，不会更改源对象，且数组会智能合并
import HtmlWebpackPlugin from 'html-webpack-plugin'

const __dirname = path.dirname(fileURLToPath(import.meta.url));

console.log(process.env.NODE_ENV)

const nextConfig = (phase, { defaultConfig }) => {
  /**
   * @type {import('next').NextConfig}
   */
  const nextConfig = {
    reactStrictMode: false,
    output: "export",
    distDir: 'dist',
    images: {
      unoptimized: true,
    },
    sassOptions: {
      additionalData: `@import "@/scss/variable.scss";`
    },
    webpack: (
      config,
      { buildId, dev, isServer, defaultLoaders, nextRuntime, webpack }
    ) => {
      if (phase === PHASE_DEVELOPMENT_SERVER) {
 
      } else if (phase === PHASE_PRODUCTION_BUILD) {

      }
      const originalHtmlWebpackPlugin = config.plugins.find(plugin => plugin instanceof HtmlWebpackPlugin);
      // console.log('originalHtmlWebpackPlugin', originalHtmlWebpackPlugin)
      // console.log(config.plugins)
      return merge(config, {
        plugins: [
          new webpack.ProvidePlugin({
            $: 'jquery',
          }),
        ]
      });
    },
  };
  return nextConfig;
};

export default nextConfig;

```

It's another fucking error

![image](https://github.com/vercel/next.js/assets/50656459/dbe38675-bb87-40de-9b14-ea2c4ae72e5f)

I'm confused.

**Comment by icyJoseph** [comment]
Hi,

`use client` doesn't mean that a component will be run only client side. It means that the module will be bundled to be send to the client. In other words, Client components are also SSR'd. 

You should maybe defer the allocation of SectionTree, so that it doesn't run in the server. I am not sure about what you are doing there, but if its top level code, accessing $ raw, then it is bound to have problems, just as accessing window/document or their APIs would cause issues.

**Reply by lovelyJason** [reply]
I need to use jquery to manipulate the dom, which was migrated from an old project.Can I only IMPORT $ from 'jquery' in the project? But  
Jquery Cannot run on the server side， it will remind me "jQuery requires a window with a document"

**Reply by icyJoseph** [reply]
Yeah, because it wants to attach it self to `window` ~ You can try to import the script in root layout, jquery, and then within a `useLayoutEffect` do the DOM manipulations. It'll complain that `useLayoutEffect` is a noop on the server, but there are ways to silence that. Try it first.

**Reply by lovelyJason** [reply]
But it does not complain about useLayoutEffect is a noop on the server

**Reply by icyJoseph** [reply]
yeah I think they have updated that, I saw in the React 19 preparation docs, https://react.dev/blog/2024/04/25/react-19-upgrade-guide#other-notable-changes:

> react-dom: Remove layout effect warning during SSR [#26395](https://github.com/facebook/react/pull/26395)

Though, does it work?

**Reply by lovelyJason** [reply]
So may I ask how to avoid the error on the server side error `$ is not defined` by the `<script src = '/js/jquery.js'/>`  in `src/app/layout.tsx`......

The project is running normally, but this error ..., oh shit

**Reply by icyJoseph** [reply]
There's two ways go about this, either you create a mental model of SSR and what's going on, or just hack it.

The problem is during that SSR pass, there's no window, and you are trying to access $ on the global scope. It'll only exist in the client, after that blocking script loads.

So you can't just make `$` references without having proper guards, similarly as to how one can't just invoke window at any level without proper guards.

Now I don't know how your code is structured, but this error would indicate that, you are either, calling $ at top level of a module, or that you are invoking a function at server render time, that invokes $. 

You have to make sure that $ is only ever invoked within exclusively client side executing code. That is useEffects, deferred callbacks, lazy loaded code (without SSR), etc. 
 
Be mindful that some of these techniques can end up in no DOM being generated during SSR, which might defeat the purpose of using Next.js at all, to some degree.

Also, to guard against potentially undefined globals, one has to do `typeof variable !== 'undefined'`, for example `typeof window !== 'undefined'`. Doing that without careful consideration can also lead to issues, which is why it is better to do this DOM manipulations within a `useLayoutEffect`

**Reply by lovelyJason** [reply]
Oh thanks!

**Reply by lovelyJason** [reply]
I understand what you said, but what I understand is that the difference between the usereffect and the uselayouteffect is not the time different? Uselayouteffect will block the browser rendering.

I see.useEffect can help us get the latest DOM, including layout size, right?

I took a look at my code, and I noticed that at the top level, there were actually references to jquery outside of effect and that's probably what caused it

**Reply by lovelyJason** [reply]
But I still don’t understand why it operates in useLayoutEffect better than in useEffect

**Reply by icyJoseph** [reply, marked as answer]
Yes that the issue then, those unchecked references to $, which is not available in the global scope of SSR

The reason why useLayoutEffect exists, is to do work on the DOM, imperatively, before the browser paints, so that you don't show the UI before doing the modifications you need.

useEffect can also see the DOM already, but then React has already yielded so that the browser can paint. Whether or not to use either, is almost a matter of business logic, is it ok for users to see the DOM before you go and do modifications to it? You might not see them now, but in production, in slower devices, or with other things happening in your application, it is possible to see this flashes of content.
